File: App/Java/app/src/main/java/com/example/tonjies/templateproject/module/select/SelectActivity.java
package com.example.tonjies.templateproject.module.select;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.View;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.common.util.ACatch;
import com.example.common.util.L;
import com.example.tonjies.templateproject.R;
import com.example.tonjies.templateproject.adapter.SelectAdapter;
import com.example.tonjies.templateproject.module.select.presenter.SelectPresenter;
import com.example.tonjies.templateproject.module.select.presenter.contract.SelectContract;
import com.example.tonjies.templateproject.module.weather.WeatherActivity;
import com.example.tonjies.templateproject.util.AStatic;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by 舍长
 * describe:城市选择Activity
 */
public class SelectActivity extends AppCompatActivity implements SelectContract.SelectView {

    //返回按钮
    @BindView(R.id.mImgBack)
    ImageView mImgBack;

    //地点标题
    @BindView(R.id.mTxtTitle)
    TextView mTxtTitle;

    //地点列表
    @BindView(R.id.mRecList)
    RecyclerView mRecList;

    //下拉刷新库，这里为了实现回弹效果
    @BindView(R.id.refreshLayout)
    SmartRefreshLayout refreshLayout;

    //布局管理器
    private LinearLayoutManager linearLayoutManager;

    //适配器
    private SelectAdapter selectAdapter;

    //列表数据
    private List<String> mData;

    //P层接口
    private SelectPresenter selectPresenter;

    //当前选中的级别
    private int currentLevel;

    //分隔线
    private DividerItemDecoration decoration;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //将背景延伸到状态栏
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
        setContentView(R.layout.activity_select);
        ButterKnife.bind(this);
        init();//初始化
    }

    /**
     * 初始化数据，控件
     */
    private void init() {

        mImgBack.setVisibility(View.GONE);
        mData = new ArrayList<>();
//        for (int i = 'A'; i < 'Z'; i++) {
//            mData.add("" + (char) i);
//        }
        //初始化recyclerView
        linearLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        mRecList.setLayoutManager(linearLayoutManager);
        selectAdapter = new SelectAdapter(this, mData);
        mRecList.setAdapter(selectAdapter);
        selectPresenter = new SelectPresenter(this);
        selectPresenter.queryFromServer("", "", AStatic.level_province);

        decoration=new DividerItemDecoration(this,DividerItemDecoration.VERTICAL);
        mRecList.addItemDecoration(decoration);
    }

    /**
     * view 就是我们点击的itemView
     */
    public void myItemClick(View view) {
        // 获取点击到d的itemView的位置
        int position = mRecList.getChildAdapterPosition(view);
        //根据不同的类型再次进行请求
        switch (currentLevel) {
            //如果是从省份列表点击的，继续查询省份下的城市数据
            case AStatic.level_province:
                //显示返回按钮
                mImgBack.setVisibility(View.VISIBLE);
                selectPresenter.queryCities(position);
                break;
            //如果是从城市列表点击的，继续查询城市下的县级数据
            case AStatic.level_city:
                mImgBack.setVisibility(View.VISIBLE);
                selectPresenter.queryCountry(position);
                break;
            //如果是从县级点击的,查询天气信息，进入主界面，结束当前Activity
            case AStatic.level_county:
                mImgBack.setVisibility(View.VISIBLE);
                selectPresenter.queryCountryWeatherId(position);
                break;
        }
        Toast.makeText(this, "点击了 " + mData.get(position),
                Toast.LENGTH_SHORT).show();
    }


    /**
     * 接受服务器返回的数
     */
    @Override
    public void setData(List<String> dataList, int type, String title) {
        linearLayoutManager.scrollToPositionWithOffset(0,0);
        mTxtTitle.setText(title);
        mData.clear();
        mData.addAll(dataList);
        selectAdapter.notifyDataSetChanged();
        currentLevel = type;
    }

    //设置天气id
    @Override
    public void setWeatherId(String weatherId) {
        L.d("setWeatherId:"+weatherId);
        ACatch.putString(AStatic.weatherId, weatherId);
        startActivity(new Intent(this,WeatherActivity.class));
    }

    //返回按钮
    @OnClick(R.id.mImgBack)
    public void onViewClicked() {
        switch (currentLevel) {
            case AStatic.level_county:
                selectPresenter.queryCities(-1);
                break;
            case AStatic.level_city:
                mImgBack.setVisibility(View.GONE);
                //查询省份信息
                selectPresenter.queryFromServer("", "", AStatic.level_province);
                break;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        selectPresenter.detachView();
    }
}

File: App/kotlins/app/src/main/java/com/example/tonjies/weatherx/module/select/SelectActivity.kt
package com.example.tonjies.weatherx.module.select

import android.app.ProgressDialog
import android.graphics.Color
import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import com.example.tonjies.weatherx.R
import com.example.tonjies.weatherx.adapter.SelectAdapter
import com.example.tonjies.weatherx.base.MvpActivity
import com.example.tonjies.weatherx.module.select.bean.City
import com.example.tonjies.weatherx.module.select.bean.County
import com.example.tonjies.weatherx.module.select.bean.Province
import com.example.tonjies.weatherx.module.select.presenter.SelectPresenter
import com.example.tonjies.weatherx.module.select.view.SelectView
import com.example.tonjies.weatherx.module.weather.WeatherActivity
import com.example.tonjies.weatherx.util.Config
import com.example.tonjies.weatherx.util.L
import com.example.tonjies.weatherx.util.ShareUtils
import kotlinx.android.synthetic.main.activity_select.*
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.toast

/**
 * 城市选择界面
 */
class SelectActivity : MvpActivity<SelectPresenter>(), SelectView {
    override fun setWeatherId(weatherId: String) {
        ShareUtils.putString(this,"weatherId",weatherId)
        startActivity<WeatherActivity>("weatherId" to weatherId)
    }

    private var layoutManger: LinearLayoutManager? = null

    private var selectAdapter = SelectAdapter()

    //请求级别
    private var LEVEL_PROVINCE = 0//省
    private var LEVEL_CITY = 1//市
    private var LEVEL_COUNT = 2//县


    //RecyclerView数据集合
    private var mDatas = ArrayList<String>()

    //当前选中的级别
    private var currentLevel: Int? = null

    //进度条
    private var progressDialog: ProgressDialog? = null

    //P层
    private var selectPresenter = SelectPresenter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //将背景延伸到状态栏
        if (Build.VERSION.SDK_INT >= 21) {
            val decorView = window.decorView
            decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
            window.statusBarColor = Color.TRANSPARENT
        }
        setContentView(R.layout.activity_select)

        //初始化本地数据
//        initData()
        //设置view
        initView()
    }

    private fun initView() {
        //设置布局管理器
        layoutManger = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)

        //设置适配器
        selectAdapter = SelectAdapter(this, mDatas)
        mRecList.layoutManager = layoutManger

        mRecList.adapter = selectAdapter
        //查询省份信息
        queryProvinces()
        mImgBack.setOnClickListener {
            if (currentLevel == LEVEL_COUNT) {
                //当选中的级别是县级
                //根据当前选中的城市id进行城市查询
                selectPresenter.queryCities(Config.provinceCode!!)
            } else if (currentLevel == LEVEL_CITY) {
                mImgBack.visibility = View.GONE
                //查询省份信息
                selectPresenter.queryFromServer("", "", "province")
            }
        }
    }

    /**
     * 查询省份信息
     */
    private fun queryProvinces() {
        mTxtTitle.text = "中国"
        mImgBack.visibility = View.GONE//隐藏返回按钮
        selectPresenter.mView = this
        selectPresenter.queryFromServer("", "", "province")
        mImgBack.visibility = View.GONE
    }

    /**
     * 初始化本地数据
     */
    private fun initData() {
        var i: Int = 'A'.toInt()
        while (i < 'Z'.toInt()) {
            mDatas.add("" + i.toChar())
            i++
        }
    }

    /**
     * 设置地点名称
     */
    override fun setData(dataList: ArrayList<String>, type: Int, title: String) {
        layoutManger!!.scrollToPositionWithOffset(0, 0)//指定RecycleView华东到指定position
        mTxtTitle.text = title//设置列表标题
        mDatas.clear()
        for (s in dataList) {
            mDatas.add(s)
        }
        currentLevel = type
        selectAdapter.notifyDataSetChanged()

    }

    override fun show(msg: String) {
        toast(msg)
    }

    /**
     * 实现RecyclerView的点击事件
     */
    fun myItemClick(view: View) {
        //获得itemView的位置
        val position = mRecList.getChildAdapterPosition(view)
        when (currentLevel) {
            //如果当前选择的级别是省份
            LEVEL_PROVINCE -> {
                //传递当前选中的省份位置，查询相应的城市
                mImgBack.visibility = View.VISIBLE//显示反复按钮
                selectPresenter.queryCities(position)
            }
            //如果当前的是城市界面
            LEVEL_CITY -> {
                mImgBack.visibility = View.VISIBLE
                selectPresenter.queryCountries(position)
            }
            //如果当前选择的级别是县
            LEVEL_COUNT -> {
                selectPresenter.queryCoutry(position)
            }
        }
//        toast("点击了" + mDatas[position])
    }

}
